vendor/src/github.com/docker/libnetwork/endpoint.go

```diff
@@ -303,6 +304,24 @@ func (ep *endpoint) isAnonymous() bool {
 	return ep.anonymous
 }
 
+// CompareAndSwap ep's serviceEnabled. If its in oldState, set it to newState
+// and return true.  If its not in oldState return false
+func (ep *endpoint) casServiceEnabled(oldState, newState bool) bool {
```

Comment:
@sanimej Yes, this only needs one method that sets the value.

Complement is implied with a boolean in CAS for this usage:

| current | old | new | set? | used
----|----|-----|-----|------
| `true` | `true` | `true` | `false` | no |
| `true` | `true` | `false` | `true` | yes |
| `true` | `false` | `false` | `false` | no |
| `true` | `false` | `true` | `false` | yes |
| `false` | `false` | `false` | `false` | no |
| `false` | `false` | `true` | `true` | yes  |
| `false` | `true` | `true` | `false` |  no |
| `false` | `true` | `false` | `false` | yes |

Pulling out the lines in the above table that represent the usage of these methods, we have the following:

| current | old | new | set? | used
----|----|-----|-----|------
| `true` | `true` | `false` | `true` | yes |
| `true` | `false` | `true` | `false` | yes |
| `false` | `false` | `true` | `true` | yes  |
| `false` | `true` | `false` | `false` | yes |

Per inspection, we can see that the return value, `set?` is independent of `old`, meaning that because you always use the complement for `old`, CAS is independent of last observed.